app/assets/v2/js/pages/profile.js

```diff
@@ -39,43 +65,146 @@ $(document).ready(function() {
-
-                activityContainer.setAttribute('page', page);
-                results.insertAdjacentHTML('afterBegin', html);
+            response.text().then(function(html) {
+              const results = document.createElement('div');
 
-                const childs = results.children;
+              activityContainer.setAttribute('page', page);
+              results.insertAdjacentHTML('afterBegin', html);
 
-                while (childs.length) {
-                  activityContainer.append(childs[0]);
-                }
+              const childs = results.children;
 
-                fetchInProgress = false;
-                loadingImg.className = loadingImg.className.replace('visible', 'hidden');
+              while (childs.length) {
+                activityContainer.append(childs[0]);
               }
-            );
+
+              fetchInProgress = false;
+              loadingImg.className = loadingImg.className.replace(
+                'visible',
+                'hidden'
+              );
+            });
             return;
           }
 
           fetchInProgress = false;
           hideBusyOverlay();
-        }
-      ).catch(
-        function() {
+        })
+        .catch(function() {
           fetchInProgress = false;
-          loadingImg.className = loadingImg.className.replace('visible', 'hidden');
-        }
-      );
+          loadingImg.className = loadingImg.className.replace(
+            'visible',
+            'hidden'
+          );
+        });
     }
   }
 
+  function toggleProfileHeaderView() {
+    banner.style['visibility'] = 'visible';
+    cancelHeaderUpdate.classList.remove('hidden');
+    updateHeader.classList.add('hidden');
+  }
+
+  function cancelProfileHeaderBannerUpdate() {
+    cancelHeaderUpdate.classList.add('hidden');
+    updateHeader.classList.remove('hidden');
+    banner.style['visibility'] = 'hidden';
+    updateHeader.innerHTML = 'Update Profile';
+  }
+
+  function setBanner() {
+    const data = new FormData();
+
+    data.append('banner', file);
+    data.append(
+      'csrfmiddlewaretoken',
+      $('input[name="csrfmiddlewaretoken"]').attr('value')
+    );
+    fetch('/api/v0.1/profile/banner', {
+      method: 'post',
+      body: data
+    }).then(response => {
+      if (response.status === 200) {
+        updateHeader.innerHTML = 'Update Profile';
+        location.reload();
+        _alert(
+          { message: gettext('User profile banner has been updated.') },
+          'success'
+        );
+      } else {
+        _alert(
+          { message: gettext('An error occured. Please try again.') },
+          'error'
+        );
+      }
+    });
+  }
+
+  function fetchSystemBanners() {
+    fetch('/api/v0.1/banners', {
+      method: 'GET'
+    }).then(response => response.json())
+      .then(response => {
+        if (response.status === 200) {
+          const content = $.parseHTML(`
+            <div class="modal fade" id="bannerUploadModal" tabindex="-1">
```

Comment:
it'd be nice to have this as a JSX template instead of hardcoded html.

Reply:
That would be nice but will have flaws on it own. Like loading and waiting for the banners to be populated. With this UX, the modal is only loaded when the banners have been fetched.